crates/span/src/ast_id.rs

```diff
@@ -4,137 +4,533 @@
+                if false {
+                    // Ensure we covered all variants.
+                    match ErasedFileAstIdKind::Root {
+                        $( ErasedFileAstIdKind::$kind => {} )*
+                    }
+                    unreachable!()
+                }
+                $( else if kind == ErasedFileAstIdKind::$kind as u32 {
+                    stringify!($kind)
+                } )*
+                else {
+                    "Unknown"
+                }
+            };
+        }
+        let kind = kind!(
+            Root,
+            Enum,
+            Struct,
+            Union,
+            ExternCrate,
+            MacroDef,
+            MacroRules,
+            Module,
+            Static,
+            Trait,
+            TraitAlias,
+            Variant,
+            Const,
+            Fn,
+            MacroCall,
+            TypeAlias,
+            ExternBlock,
+            Use,
+            Impl,
+            BlockExpr,
+            Fixup,
+        );
+        if f.alternate() {
+            write!(f, "{kind}[{:04X}, {}]", self.hash_value(), self.index())
+        } else {
+            f.debug_struct("ErasedFileAstId")
+                .field("kind", &format_args!("{kind}"))
+                .field("index", &self.index())
+                .field("hash", &format_args!("{:04X}", self.hash_value()))
+                .finish()
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
+enum ErasedFileAstIdKind {
+    Root,
+    // The following are associated with `ErasedHasNameFileAstId`.
+    Enum,
+    Struct,
+    Union,
+    ExternCrate,
+    MacroDef,
+    MacroRules,
+    Module,
+    Static,
+    Trait,
+    TraitAlias,
+    // Until here associated with `ErasedHasNameFileAstId`.
+    // The following are associated with `ErasedAssocItemFileAstId`.
+    Variant,
+    Const,
+    Fn,
+    MacroCall,
+    TypeAlias,
+    // Until here associated with `ErasedAssocItemFileAstId`.
+    // Extern blocks don't really have any identifying property unfortunately.
+    ExternBlock,
+    // FIXME: If we store the final `UseTree` instead of the top-level `Use`, we can store its name,
+    // and be way more granular for incrementality, at the expense of increased memory usage.
+    // Use IDs aren't used a lot. The main thing that stores them is the def map. So everything that
+    // uses the def map will be invalidated. That includes infers, and so is pretty bad, but our
+    // def map incrementality story is pretty bad anyway and needs to be improved (see
+    // https://rust-lang.zulipchat.com/#narrow/channel/185405-t-compiler.2Frust-analyzer/topic/.60infer.60.20queries.20and.20splitting.20.60DefMap.60).
+    // So I left this as-is for now, as the def map improvement should also mitigate this.
+    Use,
+    /// Associated with [`ImplFileAstId`].
+    Impl,
+    /// Associated with [`BlockExprFileAstId`].
+    BlockExpr,
+    /// Keep this last.
+    Fixup,
```

Comment:
Given this is kept last, that makes my previous comment even more important, as adding a new variant *will* change the discriminant for this variant